openeogeotrellis/job_tracker_v2.py

```diff
@@ -85,48 +87,112 @@ class YarnAppReportParseException(Exception):
+        #   At present we are assuming there is already a Kerberos ticket because kinit was run before
+        #   the geopyspark driver process was started.
+        #   AFAIK that means we don't have to provide user & pw which is in fact the whole point.
+
+        url = self.get_application_url(application_id=app_id)
+        response = requests.get(url, auth=self.get_authentication_provider())
+
+        # Handle error responses: We need the field "message" from "RemoteException"
+        # See https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/WebServicesIntro.html
+        #   specifically, the section: Response Errors
+        #
+        # Example of an error responses body:
+        # {
+        #     "RemoteException": {
+        #         "exception": "NotFoundException",
+        #         "message": "java.lang.Exception: app with id: application_1674538064532_46723 not found",
+        #         "javaClassName": "org.apache.hadoop.yarn.webapp.NotFoundException"
+        #     }
+        # }
+        if response.status_code == 404:
+            remote_exc = response.json().get("RemoteException", {})
+            message = remote_exc.get(
+                "message",
+                "Unknown error. Error response did not contain the message property.",
+            )
+            raise AppNotFound(message) from None
+        else:
+            # Check if there was any other HTTP error status.
+            response.raise_for_status()
+            return self.parse_application_response(json=response.json())
+
+    @staticmethod
+    def _ms_epoch_to_date(epoch_millis: int) -> Union[str, None]:
+        """Parse millisecond timestamp from app report and return as rfc3339 date (or None)"""
+        if epoch_millis == 0:
+            return None
+        utc_datetime = dt.datetime.utcfromtimestamp(epoch_millis / 1000)
+        return rfc3339.datetime(utc_datetime)
+
+    @classmethod
+    def parse_application_response(cls, json: dict) -> _JobMetadata:
+        if not json:
+            raise YarnAppReportParseException("Response is empty")
+
+        # Handle a corrupt response that comes through as a string, not valid JSON.
+        if isinstance(json, str):
+            raise YarnAppReportParseException(
+                f"Response is corrupt, not valid JSON. Received following string: {json!r}"
+            )
 
-    def parse_application_report(self, report: str) -> _JobMetadata:
-        try:
-            props = dict(re.findall(r"^\t(.+?) : (.+)$", report, flags=re.MULTILINE))
+        report = json.get("app", {})
+        required_keys = ["state", "finalStatus", "startedTime", "finishedTime"]
+        missing_keys = [k for k in required_keys if k not in report]
+        if missing_keys:
+            raise YarnAppReportParseException(
+                f"JSON response is missing following required keys: {missing_keys}"
```

Comment:
I would include the JSON response itself as well, not just the missing keys.

Reply:
Done.